crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -1495,6 +1496,364 @@ impl<'w, 'a, T: Component> VacantEntry<'w, 'a, T> {
+    #[inline]
+    pub fn components(&self) -> impl Iterator<Item = ComponentId> + '_ {
+        self.access.reads_and_writes()
+    }
+
+    /// Returns a reference to the underlying [`Access`].
+    #[inline]
+    pub fn access(&self) -> &Access<ComponentId> {
+        &self.access
+    }
+
+    /// Returns `true` if the current entity has a component of type `T`.
+    /// Otherwise, this returns `false`.
+    ///
+    /// ## Notes
+    ///
+    /// If you do not know the concrete type of a component, consider using
+    /// [`Self::contains_id`] or [`Self::contains_type_id`].
+    #[inline]
+    pub fn contains<T: Component>(&self) -> bool {
+        self.contains_type_id(TypeId::of::<T>())
+    }
+
+    /// Returns `true` if the current entity has a component identified by `component_id`.
+    /// Otherwise, this returns false.
+    ///
+    /// ## Notes
+    ///
+    /// - If you know the concrete type of the component, you should prefer [`Self::contains`].
+    /// - If you know the component's [`TypeId`] but not its [`ComponentId`], consider using
+    /// [`Self::contains_type_id`].
+    #[inline]
+    pub fn contains_id(&self, component_id: ComponentId) -> bool {
+        self.entity.contains_id(component_id)
+    }
+
+    /// Returns `true` if the current entity has a component with the type identified by `type_id`.
+    /// Otherwise, this returns false.
+    ///
+    /// ## Notes
+    ///
+    /// - If you know the concrete type of the component, you should prefer [`Self::contains`].
+    /// - If you have a [`ComponentId`] instead of a [`TypeId`], consider using [`Self::contains_id`].
+    #[inline]
+    pub fn contains_type_id(&self, type_id: TypeId) -> bool {
+        self.entity.contains_type_id(type_id)
+    }
+
+    /// Gets access to the component of type `T` for the current entity.
+    /// Returns `None` if the entity does not have a component of type `T`.
+    #[inline]
+    pub fn get<T: Component>(&self) -> Option<&'w T> {
+        let Some(id) = self.entity.world().components().get_id(TypeId::of::<T>()) else {
+            return None;
+        };
+        self.access
+            .has_read(id)
+            // SAFETY: We have read access so we must have the component
+            .then(|| unsafe { self.entity.get().debug_checked_unwrap() })
+    }
+
+    /// Gets access to the component of type `T` for the current entity,
+    /// including change detection information as a [`Ref`].
+    ///
+    /// Returns `None` if the entity does not have a component of type `T`.
+    #[inline]
+    pub fn get_ref<T: Component>(&self) -> Option<Ref<'w, T>> {
+        let Some(id) = self.entity.world().components().get_id(TypeId::of::<T>()) else {
+            return None;
+        };
+        self.access
+            .has_read(id)
+            // SAFETY: We have read access so we must have the component
+            .then(|| unsafe { self.entity.get_ref().debug_checked_unwrap() })
```

Comment:
We should document this invariant on `FilteredEntityRef::new` (and the field IMO). It's not clear right now that having a read access guarantees the component will exist on the entity.

Reply:
Very fair, this was written in the context of use in a query but it needs to be maintained in `new` as well.

Reply:
On second thought -- why use `debug_checked_unwrap` at all? We could return return `None` here.

Reply:
That's true, it was mostly about trying to avoid an extra condition when in the intended use case we don't need it.

Having to go through the ref rather than fetching the pointer directly in a standard query iteration is already more indirection in theory, although I admittedly haven't looked at the generated code to compare so don't have a strong opinion.

Reply:
I don't feel strongly but removing the unsafe, it's up to you.